blog/2021/01/snoopi_deep.md

```diff
@@ -0,0 +1,259 @@
+
+![flamegraph-complex](/assets/blog/2021-latency/flamegraph-complex.png)
+
+Here, the red boxes correspond to `MethodInstance`s which cannot be "naturally" precompiled.
+This occurs when the method is owned by one module but the argument types are from another unrelated module.
+We'll see how to deal with these in later installments.
+
+You also see breaks in the flamegraph. During these periods, code generation and runtime create new objects and then call methods on those objects; if one of those calls requires a fresh entrance into inference, that triggers the creation of a new flame.
+Hence, the number of distinct flames (which is just equal to `length(tinf.children)`) gives you a rough indication of how frequently the chains of inference were broken. This can be caused by type instability, separate top-level calls from the REPL during `@snoopi_deep`, calls to `eval` in your code, or some other cause for julia to start running type inference.
+
+### PProf.jl
+```
+julia> fg = flamegraph(tinf)
+Node(FlameGraphs.NodeData(ROOT() at typeinfer.jl:75, 0x00, 0:10080857))
+
+julia> using PProf
+
+julia> pprof(fg)
+Serving web UI on http://localhost:57599
+```
+
+For a detailed walkthrough of how to use PProf, see the [PProf.jl README](https://github.com/JuliaPerf/PProf.jl) and, for a more complete guide, the [google/pprof web interface README](https://github.com/google/pprof/blob/master/doc/README.md#web-interface).
+
+We have found the most useful views for inspecting an inference profile to be the `/flamegraph` and `/source` views. If you navigate to the Flamegraph view, via VIEW -> Flame Graph, this is what you should see:
+
+![PProf Flamegraph in full](/assets/blog/2021-latency/pprof-flamegraph-full-1.png)
+
+Of course, as above, in this example inference was only a small part of the total time. You can zoom-in on nodes by clicking on them, and hovering will display their full text and time. The nicest way to expand just the inference time is to filter out the _non-inference time_ by "hiding" the `ROOT()` node. To do this, type `ROOT()` in the `Search regexp` bar and click REFINE -> Hide, or add `?h=ROOT` to the URL. This works well if there are several top-level calls to inference, as discussed below. One benefit of hiding and filtering via the Search bar is that the percentages for all nodes will be recomputed as a percentage of the new total.
+
+Typing values into the search bar (without pressing enter) will highlight them in the graph, as displayed in the following image, and pressing enter will filter the graph to _only_ contain paths that match your search. These tools can be used to identify small functions that show up often, contributing large amounts of time when aggregated together.
+
+![PProf Flamegraph filtered and highlighted](/assets/blog/2021-latency/pprof-flamepgraph-no-root-highlight.png)
+
+Here's a screenshot of the Top view, again with `ROOT()` hidden: `http://localhost:57599/ui/top?h=ROOT`
```

Comment:
```suggestion
Here's a screenshot of the Top view, again with `ROOT()` hidden:
```

Reply:
Yeah i'm not sure how i ended up with that punctuation. I wanted to just put a tiny reminder in there that you can see what's being hidden in the URL. Maybe shrinking it down like this would help?:
```suggestion
Here's a screenshot of the Top view, again with `ROOT()` hidden (`/top?h=ROOT()`):
```